virtualenv/_utils.py

```diff
@@ -1,25 +1,34 @@
 from __future__ import absolute_import, division, print_function
 
-import errno
 import os
+import re
 import shutil
 import stat
 
 
 def ensure_directory(directory, *args, **kwargs):
-    try:
+    # Fail if the destination exists and it's not a directory
+    if not os.path.isdir(directory):
         os.makedirs(directory, *args, **kwargs)
-    except OSError as exc:
-        if exc.errno != errno.EEXIST:
-            raise
 
 
-def copyfile(srcfile, destfile):
-    # We use copyfile (not move, copy, or copy2) to be extra sure that we are
-    # not moving directories over (copyfile fails for directories) as well as
-    # to ensure that we are not copying over any metadata because we want more
-    # control over what metadata we actually copy over.
-    shutil.copyfile(srcfile, destfile)
+def copyfile(srcfile, destfile, skip=re.compile(r".*\.pyc\Z|__pycache__\Z", re.IGNORECASE)):
+    ensure_directory(os.path.dirname(destfile))
+
+    if os.path.isdir(srcfile):
+        # TODO: just use shutil.copytree to avoid bikeshedding
```

Comment:
So this seems to me very similar to what we have in Python's `shutil`. I wrote that comment to draw attention and prompt discussion about whether there's a better way to handle copying.

It could use a better name at least but I couldn't decide.